unmarshal_test.go

```diff
@@ -0,0 +1,401 @@
+		InvalidInt int `env:"INVALID_INT"`
+	}
+	err := Unmarshal(&cfg)
+	assertError(t, err, "Unmarshal InvalidInt")
+}
+
+func TestUnmarshalSetFieldFloatError(t *testing.T) {
+	setEnvForTest(t, "INVALID_FLOAT", "invalid")
+
+	var cfg struct {
+		InvalidFloat float64 `env:"INVALID_FLOAT"`
+	}
+	err := Unmarshal(&cfg)
+	assertError(t, err, "Unmarshal InvalidFloat")
+}
+
+func TestUnmarshalSetFieldNestedError(t *testing.T) {
+	type NestedConfig struct {
+		NestedField string `env:"NESTED_FIELD,required"`
+	}
+
+	setEnvForTest(t, "NESTED_FIELD", "") // Setting an empty value to trigger required error
+
+	var cfg struct {
+		Nested NestedConfig
+	}
+	err := Unmarshal(&cfg)
+	assertError(t, err, "Unmarshal NestedConfig")
+}
+
+func TestUnmarshalRequired(t *testing.T) {
+	type RequiredConfig struct {
+		RequiredVar string `env:"REQUIRED_VAR,required"`
+	}
+
+	var cfg RequiredConfig
+	err := Unmarshal(&cfg)
+	assertError(t, err, "Unmarshal Required")
+
+	expectedErr := "required environment variable REQUIRED_VAR is not set"
+	if err != nil && err.Error() != expectedErr {
+		t.Errorf("expected error %s, got %s", expectedErr, err.Error())
+	}
+
+	setEnvForTest(t, "REQUIRED_VAR", "value")
+	err = Unmarshal(&cfg)
+	assertNoError(t, err, "Unmarshal Required")
+	assertEqual(t, "value", cfg.RequiredVar, "Unmarshal Required")
+}
+
+func TestUnmarshalSetFieldErrors(t *testing.T) {
+	tests := []struct {
+		envKey    string
+		envValue  string
+		fieldType string
+	}{
+		{"INVALID_UINT", "invalid", "uint"},
+		{"INVALID_FLOAT", "invalid", "float32"},
+		{"UNSUPPORTED", "invalid", "complex64"},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.fieldType, func(t *testing.T) {
+			setEnvForTest(t, tt.envKey, tt.envValue)
+
+			var cfg struct {
+				InvalidUint  uint      `env:"INVALID_UINT"`
+				InvalidFloat float32   `env:"INVALID_FLOAT"`
+				Unsupported  complex64 `env:"UNSUPPORTED"`
+			}
+
+			err := Unmarshal(&cfg)
+			assertError(t, err, "Unmarshal "+tt.fieldType)
+		})
+	}
+}
+
+func TestSetFieldUint(t *testing.T) {
+	type Config struct {
+		UIntField uint `env:"UINT_FIELD"`
+	}
+
+	var cfg Config
+
+	err := Set("UINT_FIELD", "42")
+	assertNoError(t, err, "Set UINT_FIELD")
+
+	field := reflect.ValueOf(&cfg).Elem().FieldByName("UIntField")
+	err = setField(field, "42")
+	assertNoError(t, err, "setField Uint")
+
+	assertEqual(t, uint(42), cfg.UIntField, "UintField value")
+}
+
+func TestSetFieldBoolError(t *testing.T) {
+	type Config struct {
+		BoolField bool `env:"BOOL_FIELD"`
+	}
+
+	var cfg Config
+
+	err := Set("BOOL_FIELD", "invalid")
+	assertNoError(t, err, "Set BOOL_FIELD")
+
+	field := reflect.ValueOf(&cfg).Elem().FieldByName("BoolField")
+	err = setField(field, "invalid")
+	assertError(t, err, "setField Bool")
+}
+
+func TestUnsupportedSliceElementKind(t *testing.T) {
+	type Config struct {
+		Unsupported []complex64 `env:"UNSUPPORTED"`
+	}
+
+	var cfg Config
+
+	err := Set("UNSUPPORTED", "invalid")
+	assertNoError(t, err, "Set UNSUPPORTED")
```

Comment:
Here, and in a few other places, you could/should use the test helper


```suggestion
	setEnvForTest(t, "UNSUPPORTED", "invalid")
```